rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs

```diff
@@ -0,0 +1,874 @@
+    /// If the cursor is fully synced, this returns None.
+    /// Otherwise, it returns the next range to query, either by block or sequence depending on the mode.
+    pub async fn get_next_range(&mut self) -> Result<Option<RangeInclusive<u32>>> {
+        // Fast forward the cursor if necessary.
+        self.fast_forward().await?;
+
+        // If `self.current_indexing_snapshot` is None, we are synced and there are no more ranges to query.
+        // Otherwise, we query the next range, searching for logs prior to and including the current indexing snapshot.
+        Ok(self
+            .current_indexing_snapshot
+            .as_ref()
+            .map(|current_indexing_snapshot| match &self.index_mode {
+                IndexMode::Block => self.get_next_block_range(current_indexing_snapshot),
+                IndexMode::Sequence => self.get_next_sequence_range(current_indexing_snapshot),
+            }))
+    }
+
+    /// Gets the next block range to index.
+    /// Only used in block mode.
+    fn get_next_block_range(
+        &self,
+        current_indexing_snapshot: &TargetSnapshot,
+    ) -> RangeInclusive<u32> {
+        // Query the block range ending at the current_indexing_snapshot's at_block.
+        current_indexing_snapshot
+            .at_block
+            .saturating_sub(self.chunk_size)..=current_indexing_snapshot.at_block
+    }
+
+    /// Gets the next sequence range to index.
+    /// Only used in sequence mode.
+    fn get_next_sequence_range(
+        &self,
+        current_indexing_snapshot: &TargetSnapshot,
+    ) -> RangeInclusive<u32> {
+        // Query the sequence range ending at the current_indexing_snapshot's sequence.
+        current_indexing_snapshot
+            .sequence
+            .saturating_sub(self.chunk_size)..=current_indexing_snapshot.sequence
+    }
+
+    /// Reads the DB to check if the current indexing sequence has already been indexed,
+    /// iterating until we find a sequence that hasn't been indexed.
+    async fn fast_forward(&mut self) -> Result<()> {
+        if let Some(current_indexing_snapshot) = self.current_indexing_snapshot.clone() {
+            // Check if any new logs have been inserted into the DB,
+            // and update the cursor accordingly.
+            while self
+                .db
+                .retrieve_by_sequence(current_indexing_snapshot.sequence)
+                .await?
+                .is_some()
+            {
+                // Require the block number as well.
+                if let Some(block_number) = self
+                    .db
+                    .retrieve_log_block_number_by_sequence(current_indexing_snapshot.sequence)
+                    .await?
+                {
+                    self.last_indexed_snapshot = LastIndexedSnapshot {
+                        sequence: Some(current_indexing_snapshot.sequence),
+                        at_block: block_number.try_into()?,
+                    };
```

Comment:
agree there's room for deduping stuff between the two. I actually intentionally avoided trying to share logic because this felt like a source of confusion & bugs in the past implementation

Reply:
fair